text/0053-helpers.md

```diff
@@ -0,0 +1,327 @@
+
+```hbs
+{{my-login-button isAdmin=(can-access 'admin')}}
+```
+
+Let's step through exactly what happens when using an helper like this:
+
+```hbs
+<my-login-button isAdmin={{can-access 'admin'}} />
+```
+
+Upon initial render:
+
+* The helper `can-access` is looked up on the container
+* The helper is identified as a full helper, not a shorthand helper function
+* The helper is initialized (`init` is called)
+* The `compute` function is called on the helper.
+* The return value from `compute` is passed as an `attr` to `my-login-button`.
+* The helper instance remains in memory.
+
+If the parent scope is rerendered:
+
+* The `compute` function is called again.
+* The return value from `compute` is passed as an `attr` to `my-login-button`.
+
+Upon teardown:
+
+* The helper is destroyed, calling the `destroy` method.
+
+### Returning a value
+
+The return value of helper is passed through to where their subexpression
+is called. For example, given a helper (this one a shorthand helper):
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
+
+export default Ember.Helper.helper(function fullName(params, hash) {
+  return params.join(' ');
+}
+```
+
+The following are effectively the same:
+
+```hbs
+<div data-name={{full-name "Fenton" "Crackshell"}}></div>
+<div data-name={{"Fenton Crackshell"}}></div>
+```
+
+```hbs
+{{my-component name=(full-name "Magica" "De Spell")}}
+{{my-component name="Magica De Spell"}}
+```
+
+```hbs
+<p>{{full-name "Bentina" "Beakley"}}</p>
+<p>{{"Bentina Beakley"}}</p>
+```
+
+An exclusion to this pattern is the following form:
+
+```hbs
+<div {{full-name "Webbigail" "Vanderquack"}}></div>
+```
+
+This is a legacy form of mustache usage. Helpers will throw an exception when
+used in this manner.
+
+### Consuming services and recompute
+
+Helpers are a valid target for service injection. For example:
+
+```js
+// app/helpers/current-user-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  // Same API as components:
+  session: Ember.inject.service(),
+  compute() {
+    return this.get('session.currentUser.name');
+  }
+});
+```
+
+However consuming a property from a service does not bind the data being
+displayed to that property. After `{{current-user-name}}` has been computed
+and rendered, it will never be invalidated.
+
+For this reason, helpers are granted some control over their
+computation lifecycle. A helper will recompute when:
+
+* A value passed via the template changes (`params` or `hash`)
+* The `recompute` method is called
+
+For example, this helper checks if the current use has access to a
+resource type:
+
+```js
+// app/helpers/can-access.js
+import Ember from "ember";
+
+// Usage {{if (can-access 'admin') 'Welcome, boss' 'Heck no!'}}
+export default Ember.Helper.extend({
+  session: Ember.inject.service(),
+  onCurrentUserChange: Ember.observes('session.currentUser', function() {
+    this.recompute();
+  }),
```

Comment:
I would like to see what cases are common before adding sugar, especially since ES7 decorators will shortly make many of these things much nicer to express.